experimental/packages/sampler-composite/src/traceidratio.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ComposableSampler, SamplingIntent } from './types';
+import { INVALID_THRESHOLD, MAX_THRESHOLD } from './util';
+import { serializeTh } from './tracestate';
+
+export class ComposableTraceIDRatioBasedSampler implements ComposableSampler {
+  private readonly intent: SamplingIntent;
+  private readonly description: string;
+
+  constructor(ratio: number) {
+    if (ratio < 0 || ratio > 1) {
+      throw new Error(
+        `Invalid sampling probability: ${ratio}. Must be between 0 and 1.`
+      );
+    }
+    const threshold = calculateThreshold(ratio);
+    const thresholdStr =
+      threshold === MAX_THRESHOLD ? 'max' : serializeTh(threshold);
+    if (threshold !== MAX_THRESHOLD) {
+      this.intent = {
+        threshold: threshold,
+        thresholdReliable: true,
+      };
+    } else {
+      // Same as AlwaysOff, notably the threshold is not considered reliable
```

Comment:
While the spec says "an instance of AlwaysOff" should be returned, since it seems JS uses classes more than factory functions, I went with just reimplementing AlwaysOff here. It makes the decision more sane anyways i.e. I disagree with that spec wording